DataFormats/Detectors/CTP/src/CTPRateFetcher.cxx

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "DataFormatsCTP/CTPRateFetcher.h"
+
+#include <map>
+#include <vector>
+#include "CommonConstants/LHCConstants.h"
+
+using namespace o2::ctp;
+double CTPRateFetcher::fetch(o2::ccdb::BasicCCDBManager* ccdb, uint64_t timeStamp, int runNumber, std::string sourceName)
+{
+  auto triggerRate = fetchNoPuCorr(ccdb, timeStamp, runNumber, sourceName);
+  if (triggerRate >= 0) {
+    return pileUpCorrection(triggerRate);
+  }
+  return -1;
+}
+double CTPRateFetcher::fetchNoPuCorr(o2::ccdb::BasicCCDBManager* ccdb, uint64_t timeStamp, int runNumber, std::string sourceName)
+{
+  setupRun(runNumber, ccdb, timeStamp, 1);
+  if (sourceName.find("ZNC") != std::string::npos) {
+    if (runNumber < 544448) {
+      return fetchCTPratesInputsNoPuCorr(timeStamp, 25) / (sourceName.find("hadronic") != std::string::npos ? 28. : 1.);
+    } else {
+      return fetchCTPratesClassesNoPuCorr(timeStamp, "C1ZNC-B-NOPF-CRU", 6) / (sourceName.find("hadronic") != std::string::npos ? 28. : 1.);
+    }
+  } else if (sourceName == "T0CE") {
+    return fetchCTPratesClassesNoPuCorr(timeStamp, "CMTVXTCE-B-NOPF");
+  } else if (sourceName == "T0SC") {
+    return fetchCTPratesClassesNoPuCorr(timeStamp, "CMTVXTSC-B-NOPF");
+  } else if (sourceName == "T0VTX") {
+    if (runNumber < 534202) {
+      return fetchCTPratesClassesNoPuCorr(timeStamp, "minbias_TVX_L0", 3); // 2022
+    } else {
+      double_t ret = fetchCTPratesClassesNoPuCorr(timeStamp, "CMTVX-B-NOPF");
+      if (ret == -2.) {
+        LOG(info) << "Trying different class";
+        ret = fetchCTPratesClassesNoPuCorr(timeStamp, "CMTVX-NONE");
+        if (ret < 0) {
+          LOG(fatal) << "None of the classes used for lumi found";
+        }
+      }
+      return ret;
+    }
+  }
+  LOG(error) << "CTP rate for " << sourceName << " not available";
+  return -1.;
+}
+void CTPRateFetcher::updateScalers(ctp::CTPRunScalers& scalers)
+{
+  mScalers = scalers;
+  mScalers.convertRawToO2();
+}
+//
+double CTPRateFetcher::fetchCTPratesClasses(uint64_t timeStamp, const std::string& className, int inputType)
+{
+  auto triggerRate = fetchCTPratesClassesNoPuCorr(timeStamp, className, inputType);
+  if (triggerRate >= 0) {
+    return pileUpCorrection(triggerRate);
+  }
+  return -1;
+}
+double CTPRateFetcher::fetchCTPratesClassesNoPuCorr(uint64_t timeStamp, const std::string& className, int inputType)
+{
+  std::vector<ctp::CTPClass> ctpcls = mConfig.getCTPClasses();
```

Comment:
Do you really need a copy of the vector? Would not const reference suffice?
Also the next vector.

Reply:
done